src/telemetry/writer.cpp

```diff
@@ -0,0 +1,233 @@
+
+    rjwriter_.EndArray();
+  }
+
+  // crucial data points that are displayed in various fixed GUI points
+  // NOT EDITABLE
+  void Writer::packCrucialData()
+  {
+    rjwriter_.Key("crucial_data");
+    rjwriter_.StartArray();
+
+    data::Navigation nav_data = data_.getNavigationData();
+    data::StateMachine sm_data = data_.getStateMachineData();
+    add("distance", 0.0, 1250.0, "m", nav_data.distance);
+    add("velocity", 0.0, 250.0, "m/s", nav_data.velocity);
+    add("acceleration", -50.0, 50.0, "m/s^2", nav_data.acceleration);
+    add("status", sm_data.current_state);
+
+    rjwriter_.EndArray();
+  }
+
+  // status data points that are displayed in the GUI status tab
+  // NOT EDITABLE
+  void Writer::packStatusData()
+  {
+    rjwriter_.Key("status_data");
+    rjwriter_.StartArray();
+
+    // TODO(everyone): add all required data points
+
+    rjwriter_.EndArray();
+  }
+
+  Writer::Writer(data::Data& data)
+    : rjwriter_(sb_),
+      data_ {data}
+  {}
+
+  void Writer::start()
+  {
+    rjwriter_.StartObject();
+  }
+
+  void Writer::end()
+  {
+    rjwriter_.EndObject();
+  }
+
+  void Writer::startList(std::string name)
+  {
+    rjwriter_.StartObject();
+    rjwriter_.Key("name");
+    rjwriter_.String(name.c_str());
+    rjwriter_.Key("value");
+    rjwriter_.StartArray();
+  }
+
+  void Writer::endList()
+  {
+    rjwriter_.EndArray();
+    rjwriter_.EndObject();
+  }
+
+  std::string Writer::getString()
+  {
+    return sb_.GetString();
+  }
+
+  void Writer::add(std::string name, int min, int max, std::string unit, int value)
+  {
+    rjwriter_.StartObject();
+    rjwriter_.Key("name");
+    rjwriter_.String(name.c_str());
+    rjwriter_.Key("min");
+    rjwriter_.Int(min);
+    rjwriter_.Key("max");
+    rjwriter_.Int(max);
+    rjwriter_.Key("unit");
+    rjwriter_.String(unit.c_str());
+    rjwriter_.Key("value");
+    rjwriter_.Int(value);
+    rjwriter_.EndObject();
+  }
+
+  void Writer::add(std::string name, float min, float max, std::string unit, float value)
+  {
+    rjwriter_.StartObject();
+    rjwriter_.Key("name");
+    rjwriter_.String(name.c_str());
+    rjwriter_.Key("min");
+    rjwriter_.Double(min);
+    rjwriter_.Key("max");
+    rjwriter_.Double(max);
+    rjwriter_.Key("unit");
+    rjwriter_.String(unit.c_str());
+    rjwriter_.Key("value");
+    rjwriter_.Double(value);
+    rjwriter_.EndObject();
+  }
+
+  void Writer::add(std::string name, bool value)
+  {
+    rjwriter_.StartObject();
+    rjwriter_.Key("name");
+    rjwriter_.String(name.c_str());
+    rjwriter_.Key("value");
+    rjwriter_.Bool(value);
+    rjwriter_.EndObject();
+  }
+
+  void Writer::add(std::string name, data::State value)
+  {
+    rjwriter_.StartObject();
+    rjwriter_.Key("name");
+    rjwriter_.String(name.c_str());
+    rjwriter_.Key("value");
+    rjwriter_.String(convertStateMachineState(value).c_str());
+    rjwriter_.EndObject();
+  }
+
+  void Writer::add(std::string name, data::ModuleStatus value)
```

Comment:
You always pass these function a `const char*` for `name` argument. Can you change the signature to avoid conversion to/from string? I see you do not use string at all but get the `name.c_str()` anyway.

Just change this to `void Writer::add(const char* name, data::ModuleStatus value)`. And similarly for the other `add()` functions.